README.md
[![Unit tests](https://github.com/hnesk/browse-ocrd/workflows/Unit%20tests/badge.svg?branch=master)](https://github.com/hnesk/browse-ocrd/actions/workflows/unittest.yml)
[![Docker build](https://github.com/hnesk/browse-ocrd/actions/workflows/dockerhub.yml/badge.svg)](https://github.com/hnesk/browse-ocrd/actions/workflows/dockerhub.yml)
[![PyPI version](https://badge.fury.io/py/browse-ocrd.svg)](https://badge.fury.io/py/browse-ocrd)
# OCR-D Browser

An extensible viewer for [OCR-D](https://ocr-d.de/) [mets.xml](https://ocr-d.de/en/spec/mets) files

 * [Screenshot](#screenshot)
 * [Features](#features)
 * [Installation](#installation)
    * [Native](#native-tested-on-ubuntu-18042004)
       * [From source](#from-source)
       * [Via pip](#via-pip)
    * [Docker](#docker)
 * [Usage](#usage)
    * [Native GUI](#native-gui)
    * [Docker service](#docker-service)
 * [Configuration](#configuration)
    * [Configuration file locations](#configuration-file-locations)
    * [Configuration file syntax](#configuration-file-syntax)
 
## Screenshot

![OCRD Browser with Page and Xml view](docs/screenshot.png)


## Features

- Browse fileGrps and pages, arranging views next to each other for comparison
- PageView: Show original or derived page images with [PAGE-XML](https://ocr-d.de/en/spec/page) annotations overlay, similar to [PageViewer](https://github.com/PRImA-Research-Lab/prima-page-viewer)
- ImageView: Show original or derived images (`AlternativeImage` on any level of the structural hierarchy)
- ImageView: Show multiple images at once for different pages (horizontally) or different segments (vertically), zooming freely
- XmlView: Show raw [PAGE-XML](https://ocr-d.de/en/spec/page) with syntax highlighting, open with [PageViewer](https://github.com/PRImA-Research-Lab/prima-page-viewer)
- TextView: Show concatenated [PAGE-XML](https://ocr-d.de/en/spec/page) text annotation
- DiffView: Show a simple diff comparison between text annotations from different fileGrps  
- HtmlView: Show rendered HTML comparison from [dinglehopper](https://github.com/qurator-spk/dinglehopper) evaluations

## Installation

OCR-D Browser requires Python 3.7 or higher.

### Native (tested on Ubuntu 18.04/20.04) 

The native installation requires [GTK 3](https://www.gtk.org/).

In any case you need a [virtual environment](https://packaging.python.org/tutorials/installing-packages/#creating-virtual-environments) with a current `pip` version (>=20), preferably your existing OCR-D venv:

<details>
  <summary>Create a current pip venv:</summary>

```bash
sudo apt install python3-pip python3-venv 
python3 -m venv venv
source venv/bin/activate
pip install --upgrade pip setuptools wheel
```
</details>


#### From source
```bash
git clone https://github.com/hnesk/browse-ocrd.git 
cd browse-ocrd
sudo make deps-ubuntu
make install
```

#### Via pip

```bash
sudo apt install libcairo2-dev libgirepository1.0-dev
pip install browse-ocrd
```

### Docker

If you have installed [Docker](https://docs.docker.com/get-docker/), you can build OCR-D Browser as a **web service**:

    docker build -t ocrd_browser .

Or use a prebuilt image from Dockerhub:

    docker pull hnesk/ocrd_browser


## Usage

### Native GUI
Start the app with the filesystem path to the METS file of your [OCR-D workspace](https://ocr-d.de/en/spec/glossary#workspace):
```
browse-ocrd ./path/to/mets.xml
```

You can still open another METS file from the UI though.

### Docker service

When running the webservice, you need to pass a directory `DATADIR` which (recursively) contains all the workspaces you want to serve.
The top entrypoint `http://localhost/` will show an index page with a link `http://localhost/browse/...` for each workspace path.
Each link will run `browse-ocrd` at that workspace in the background, and then redirect your browser to the internal [Broadway server](https://docs.gtk.org/gtk3/broadway.html), which renders the app in the web browser.

To start up, just do:

    docker run -it --rm -v DATADIR:/data -p 8085:8085 -p 8080:8080 ocrd_browser


## Configuration

### Configuration file locations

At startup the following directories a searched for a config file named `ocrd-browser.conf` 

```python
# directories and their default values under Ubuntu 20.04
GLib.get_system_config_dirs()  # '/etc/xdg/xdg-ubuntu/ocrd-browser.conf', '/etc/xdg/ocrd-browser.conf'
GLib.get_user_config_dir()     # '/home/jk/.config/ocrd-browser.conf'  
os.getcwd()                    # './ocrd-browser.conf'
```

### Configuration file syntax

The `ocrd-browser.conf` file is an ini-file with the following sections and keys:
```ini
[FileGroups]
# Preferred fileGrp names for thumbnail display in the Page Browser 
# Comma separated list of regular expressions
preferredImages = OCR-D-IMG, OCR-D-IMG.*, ORIGINAL

# Each Tool has a section header [Tool XYZ]
# At the moment the only defined tool is "PageViewer"  
[Tool PageViewer]
# shell commandline to execute with placeholders  
commandline = /usr/bin/java -jar /home/jk/bin/JPageViewer/JPageViewer.jar --resolve-dir {workspace.directory} {file.path.absolute}
```

> Note: You can get PRImA's PageViewer at [Github](https://github.com/PRImA-Research-Lab/prima-page-viewer/releases).


The `commandline` string will be used as a python format string with the keyword arguments:

* `workspace` : The current `ocrd.Workspace`, all properties get shell escaped (by `shlex.quote`) automatically.
* `file` : The current `ocrd_models.OcrdFile`, all properties get shell escaped (by `shlex.quote`) automatically, also there is an additional property `path` with the properties `absolute` and `relative`, so `{file.path.absolute}` will be replaced by the shell quoted absolute path of the file. 

### Configuration by environment variables

It is possible to set or override values of the configuration through environment variables. The environment variables follow this structure :  `BROCRD__{SECTION}__{KEY}`, where `SECTION` and `KEY` are in upper snake case and divided by a double underscore (`__`). If the section title contains spaces, the single words are also divided by `__`.  

Some examples:
```shell
BROCRD__FILE_GROUPS__PREFERRED_IMAGES='THUMB'  
BROCRD__TOOL__PAGEVIEWER__COMMANDLINE='ls {file.path.absolute}'  

```



CHANGELOG.md
# Change Log

<!-- 
## [Unreleased](../../compare/v0.5.5...master)
-->

## [0.5.5](../../compare/v0.5.4...0.5.5)

Features:
 * Fullscreen mode via F11
 * add CLI options for better Dockerization [#61](../../issues/61): 
   * `-m`, `--maximize` Open in maximized window
   * `-f`, `--fullscreen` Open in fullscreen window
   * `-r`, `--restricted` Open restricted (no edit/open functionality)
   * `-v`, `--version` commandline flag

Changed:
 * Refactored configuration to use pydantic, allows for overriding configurations settings by environment, e.g.:

   `BROCRD__FILE_GROUPS__PREFERRED_IMAGES='THUMB,IMG.*'`

Fixed:
 * Use last fileGrp as default (hoping that it is the fileGrp with the richest annotations ) [#60](../../issues/60) 
 * Application crashes on launch [#62](../..issues/62)

   Added the generated file ui.gresource to git. Previously it was not possible to install via `pip install git+https://github.com/hnesk/browse-ocrd` which is a pity.
 
## [0.5.4](../../compare/v0.5.3...0.5.4)

Changed: 
 * build via python -m build and setup.cfg

Fixed:
 *  False warning about number of images per grp/page [#51](../../issues/51)
 *  adapt to new import.metadata.entry_points API [#53](../../pull/53), thanks @bertsky
 *  Dockerfile: use local pkg, not PyPI [#56](../../pull/56), thanks @bertsky
 *  AttributeError: 'EntryPoints' object has no attribute 'get' [#57](../../issues/57) 
 *  Document.reorder: invalidate OcrdMets cache before save [#58](../../pull/58), thanks @kba

## [0.5.3](../../compare/v0.5.2...v0.5.3) - 2022-07-19

Breaking: 
 * Make explicit: requires now python>=3.7 (v0.5.1 is the last supported python3.6 version)

Changed:
 * Startup time improved by not using pkg_resources [#48](../../pull/48)  


## [0.5.2](../../compare/v0.5.1...v0.5.2) - 2022-06-10

New Features:
 * Automatically build a docker image at [`hnesk/ocrd_browser`](https://hub.docker.com/r/hnesk/ocrd_browser), thanks @bertsky 

Fixed:
 * PageView: improve reading order and region rendering order #44 [#44](../../issues/44)
 * Use `ocrd_utils.getLogger` instead `logging.getLogger` 

Changed:
 * Performance improvements 
   * use [fastentrypoints](https://github.com/ninjaaron/fast-entry_points)
   * use annotations
   * reduce imports with `if TYPE_CHECKING` 
 * Use [codespell](https://github.com/codespell-project/codespell) in CI
 * Stop supporting python-3.6, start supporting 3.9

## [0.5.1](../../compare/v0.5...v0.5.1) - 2022-05-10

New Features:
 * PageView: add baselines if available  [#34](../../issues/34)
 * PageView: add screenshot button  [#41](../../pull/41), thanks @bertsky
 * PageView: show @conf in tooltip [#39](../../pull/39), thanks @bertsky
 * Text/Xml/Diff view: Allow zooming with CTRL+mousewheel [#33](../../pull/33), thanks @bertsky

Fixed:
 * PageView: component menu not editable with older PAGE-XML Namespaces [#35](../../issues/35)
 * support path names with spaces [#40](../../issues/40)
 * ViewPage: ignore AlternativeImage if not retrievable [#37](../../issues/37)


## [0.5](../../compare/v0.4.3...v0.5) - 2021-11-09

New Features:
* [Added a PageView](../../pull/30) ([#15](../../issues/15) / [#28](../../issues/28)) \
  Shows page image overlaid with [PAGE-XML](https://ocr-d.de/en/spec/page) annotations, similar to PRImA-[PageViewer](https://github.com/PRImA-Research-Lab/prima-page-viewer)
  * Selectable `<Page>` or any `<AlternativeImage>` image as a base   
  * Selectable features as overlay: border, printSpace, region order, regions, lines, words, glyphs, warnings
  * Warnings: Display regions with problematic coordinates in red
  * Tooltip with current coordinates, region, text annotations

* ImageView:
  * zoom keyboard shortcuts `<CTRL>+`/`<CTRL>-`/`<CTRL>0`/`<CTRL>#`/`<CTRL><ALT>#`

Changed:
* Performance / responsiveness improvements 
  * Events get queued 
  * Introduction of `LazyPage` a lazy loading page proxy


## [0.4.3](../../compare/v0.4.2...v0.4.3) - 2021-07-22 

Fixed: 
* require gi >= 3.28 [#26](../../pull/26)

Changed:
* [Using Github actions for CI](https://github.com/hnesk/browse-ocrd/actions/workflows/unittest.yml)
* Get rid of DEFAULT_FILE_GROUP = 'OCR-D-IMG'
* Silence xsd error messages on stderr, when reading files produced by ocrd-cis-align 

New Features:
* Added a Diff-View [#13](../../issues/13) & [#29](../../pull/29)

## [0.4.2](../../compare/v0.4.1...v0.4.2) - 2020-11-12 

Fixed: 

* Catch empty imageFilename case &  don't download remote urls upfront [#25](../../issues/25)  
* Pillow workaround: convert 16-bit images to 8-bit [#23](../../pull/23)
* Don't crash on unhandled mimetype [#18](../../issues/18)
* typing.OrderedDict is not available in python 3.6  [#22](../../issues/22)

New Features

* Views can be resized and split now [#12](../../issues/12) 
* Added Webkit-HTML-View for [dinglehopper](https://github.com/qurator-spk/dinglehopper) [#25](../../pull/25) (partially fixes [#13](../../issues/13))
* Added a button to launch [PageViewer](https://www.primaresearch.org/tools/PAGEViewer) from XmlView [#21](../../pull/21) 


## [0.4.1](../../compare/v0.4.0...v0.4.1) - 2020-10-30

Fixed:

* Fix segfault: Dont put scrollable Gtk.Viewports in scrollable Gtk.Sourceviews / Removes self.viewport from View / Fixes #17
* chdir to workspace directory, fixes #19

## [0.4.0](../../compare/v0.3.0...v0.4.0) - 2020-10-22

Changed:

* Introduce Zoom #14
* show all derived images per page #14
* add Text view (simple concatenated) #10 
  
Fixed:
*  Support conversion of PIL LA image via cv #5
*  adapt to ocrd>=2.17 logging #6 

## [0.3.0](../../compare/v0.1.9...v0.3.0) - 2020-08-13

A lot:
* Found a basic structure 
* Started unittests
* Implement [scanning](https://github.com/hnesk/browse-ocrd-physical-import) as an extension

## [0.1.9] - 2020-08-01

First more or less public release 





serve.py
import os
import sys
import io
from functools import lru_cache
from pathlib import Path
from subprocess import Popen
from shutil import which
import html
from http import HTTPStatus
from http.server import SimpleHTTPRequestHandler
from socketserver import ForkingTCPServer
import urllib
import click

class RequestHandler(SimpleHTTPRequestHandler):
    bwhost = "localhost"
    bwport = 8085
    basedir = "."

    def do_GET(self):
        # FIXME: HTTP auth # self.server.auth / self.checkAuthentication()
        self.bwhost = self.headers.get('Host').split(':')[0]
        # serve a listing of available workspaces with links to ocrd-browse them
        if self.path == '/':
            self._serve_workspaces()
        # serve a single workspace
        elif self.path.startswith('/browse/'):
            path = self.path[7:].lstrip('/')
            path = urllib.parse.unquote(path)
            path = os.path.join(self.basedir, path)
            self._browse_workspace(path)
        elif self.path == '/reindex':
            self._workspaces.cache_clear()
            self.send_response(HTTPStatus.OK)
        else:
            #SimpleHTTPRequestHandler.do_GET(self)
            self.send_response(HTTPStatus.FORBIDDEN)

    def do_POST(self):
        self.do_GET()

    @property
    @lru_cache(maxsize=1)
    def _workspaces(self):
        # recursively find METS file paths
        paths = []
        for mets in Path(self.basedir).rglob('mets.xml'):
            if mets.match('.backup/*/mets.xml'):
                continue
            print(mets)
            paths.append(str(mets))
        return paths

    def _serve_workspaces(self):
        # generate HTML with links
        enc = sys.getfilesystemencoding()
        basedir = os.path.realpath(self.basedir)
        title = html.escape(basedir, quote=False)
        title = 'OCR-D Browser for workspaces at %s' % title
        r = []
        r.append('<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" '
                 '"http://www.w3.org/TR/html4/strict.dtd">')
        r.append('<html>\n<head>')
        r.append('<meta http-equiv="Content-Type" '
                 'content="text/html; charset=%s">' % enc)
        r.append('<title>%s</title>\n</head>' % title)
        r.append('<body>\n<h1>%s</h1>' % title)
        r.append('<hr>\n<ul>')
        for name in self._workspaces:
            relname = os.path.relpath(name, basedir)
            linkname = os.path.join('/browse', relname)
            r.append('<li><a href="%s">%s</a></li>'
                     % (urllib.parse.quote(linkname, errors='surrogatepass'),
                        html.escape(relname, quote=False)))
        r.append('</ul>\n<hr>\n</body>\n</html>\n')
        # write to file
        encoded = '\n'.join(r).encode(enc, 'surrogateescape')
        f = io.BytesIO()
        f.write(encoded)
        f.seek(0)
        # send file
        self.send_response(HTTPStatus.OK)
        self.send_header("Content-type", "text/html; charset=%s" % enc)
        self.send_header("Content-Length", str(len(encoded)))
        self.end_headers()
        self.copyfile(f, self.wfile)

    def _browse_workspace(self, path):
        if not os.path.exists(path):
            self.send_response(HTTPStatus.NOT_FOUND)
        else:
            if not path.endswith('mets.xml'):
                path = os.path.join(path, 'mets.xml')
            ## run app
            Popen([which('browse-ocrd'),
                         '--display', ':' + str(self.bwport - 8080),
                         path])
            ## proxy does not work, because the follow-up requests would need to be forwarded, too:
            # response = urllib.request.urlopen('http://' + self.bwhost + ':' + str(self.bwport))
            # self.send_response(response.status)
            # for name, value in response.getheaders():
            #     self.send_header(name.rstrip(':'), value)
            # self.end_headers()
            # self.copyfile(response, self.wfile)
            ## so let's use temporary redirect instead
            self.send_response(HTTPStatus.SEE_OTHER) # or TEMPORARY_REDIRECT?
            self.send_header('Location', 'http://' + self.bwhost + ':' + str(self.bwport))
            self.end_headers()

    # def log_message(self, format, *args):
    #     """ Override to prevent stdout on requests """
    #     pass

@click.command(context_settings={'help_option_names': ['-h', '--help']})
@click.option('-P', '--bwport', default=8085, type=int, help="TCP port of Broadwayd to delegate to")
@click.option('-p', '--port', default=80, type=int, help="TCP port to bind the web server to")
@click.option('-l', '--listen', default="", type=str, help="network address to bind the web server to")
@click.option('-d', '--basedir', default=".", type=click.Path(exists=True, file_okay=False))                    
def cli(bwport, port, listen, basedir):
    Handler = RequestHandler
    Handler.bwport = bwport
    Handler.basedir = basedir
    ForkingTCPServer.allow_reuse_address = True
    with ForkingTCPServer((listen, port), Handler) as httpd:
        httpd.serve_forever()

if __name__ == '__main__':
    cli()
